keps/sig-node/127-user-namespaces/README.md

```diff
@@ -316,121 +325,121 @@ The picked range will be stored under a file named `userns` in the pod folder
-volume SetUpAt() function to pass this new param when calling AtomicWriter.
+This means user 0 inside the container is mapped to user 65536 on the host. Let'
+say the pod is running UID 0 inside the container too.
 
-We have written already this code and the per-volume diff is 3 lines.
+Let's say also that this pod has a configmap with a file named `foo`.
+The file `/var/lib/kubelet/.../volumes/configmap/foo` is created by the kubelet,
+the owner is host UID 0, and bind-mounted into the container at
+`/vol/configmap/foo`.
 
-For fsGroup() we don't need any changes, these volume types already honors it.
+##### Example without idmap mounts
 
-[volume-mounter-args]: https://github.com/kubernetes/kubernetes/blob/cfd69463deeebfc5ae8a0813d7d2b125033c4aca/pkg/volume/volume.go#L125-L129
-[operation-executor]: https://github.com/kubernetes/kubernetes/blob/cfd69463deeebfc5ae8a0813d7d2b125033c4aca/pkg/volume/util/operationexecutor/operation_generator.go#L674-L675
-[volumeHost interface]: https://github.com/kubernetes/kubernetes/blob/cfd69463deeebfc5ae8a0813d7d2b125033c4aca/pkg/volume/plugins.go#L360-L361
-[atomic-writer-fsUser]: https://github.com/kubernetes/kubernetes/blob/cfd69463deeebfc5ae8a0813d7d2b125033c4aca/pkg/volume/util/atomic_writer.go#L68
-[configmap-fsuser]: https://github.com/kubernetes/kubernetes/blob/cfd69463deeebfc5ae8a0813d7d2b125033c4aca/pkg/volume/configmap/configmap.go#L272-L273
+To understand how idmap mounts work, let's first see an example of what happens
+without idmap mounts.
 
-#### Phase 2: pods with volumes
+If the pod wants to read who is the owner of file `/vol/configmap/foo`, as host
+UID 0 is not mapped, it will be shown as nobody. Therefore, it won't be able to
+read the file (unless it has permissions for others).
 
-This phase makes user namespaces work in pods with volumes too. This is
-activated via the bool `pod.spec.HostUsers` too and pods fall into this mode if
-some other volume type than the ones listed for phase 1 is used. IOW, when phase
-2 is implemented, pods that use volume types not supported in phase 1, fall into
-the phase 2.
+Therefore, if we want the pod to be able to read those files and permissions be
+what we expect, the kubelet needs to chown the file to UID 65536, as that is
+what root inside the container is mapped to in the user namespaces.
 
-All pods in this mode will use _the same_ mapping, chosen by the kubelet, with a
-length 65536, and mapping the range 0-65535 too. IOW, each pod will have its own user
-namespace, but they will map to _the same_ UIDs/GIDs in the host.
+We tried this before, but several limitations were hit:
 
-Using the same mapping allows for pods to share files and mitigates all the
-listed vulnerabilities (as the host is protected from the container). It is also
-a trivial next-step to take, given that we have phase 1 implemented: just return
-the same mapping if the pod has other volumes.
+* Changing the owner of files in configmaps/secrets/etc was a concern, those
```

Comment:
Could link to the discussion or list the concerns here.

Reply:
Added, thanks!

Reply:
Note this is now in the "Alternatives" section